### ISSUE TITLE
Implement lens equation solver

### ISSUE THREAD

**Issue by adam-coogan** [Ciela-Institute/caustic#6]
This is needed for determining the positions of images of point-like sources (e.g. quasars). I'm imagining something like this:
```python
from typing import Union
from caustic.lenses.base import AbstractThinLens, AbstractThickLens

def forward_raytrace(
    thx0_src,
    thy0_src,
    lens: Union[AbstractThinLens, AbstractThickLens],
    z_l,
    z_s,
    cosmology,
    *args_lens,
    **kwargs_lens,
):
    def fn(thx, thy):
        """
        Function to minimize: distance between reverse ray-traced points and source's
        position.
        """
        bx, by = lens.raytrace(thx, thy, z_l, z_s, cosmology, *args_lens, **kwargs_lens)
        return (bx - thx0_src)**2 + (by - thy0_src)**2
    
    # TODO: implement. Could get guesses for image positions and fit from there.
    ...
```
It'd be good to write this as a standalone function first and then think about turning it into a method for the base lens classes.

For reference, Lenstronomy handles this in the [LensEquationSolver](https://lenstronomy.readthedocs.io/en/latest/lenstronomy.LensModel.Solver.html) class, which contains a generic solver and a faster one that works with EPL + shear models only. Ideally we'd implement both approaches.

**Comment by adam-coogan** [comment]
This is roughly working! The main issue is that [levmarq-torch](https://github.com/adam-coogan/levmarq-torch) is not batched very well. It also needs to be added as a dependency.

**Comment by ConnorStoneAstro** [comment]
Closed by #67
